File: cpp/session/CommandCounterTest.cpp
#include "se/StandardCommandType.hpp"
#include "session/CommandCounter.hpp"
#include "session/Timer.hpp"

#include "common/Test.hpp"
#include "common/init.hpp"

using namespace std;

void testLoginCount()
{
    Timer::setTime("20070101.010101");
    CommandCounter cc(5000);
    ASSERT_EQ(0, cc.getTotal());
    ASSERT_EQ(0, cc.getCount(StandardCommandType::LOGIN()));

    cc.increment(StandardCommandType::LOGIN());
    ASSERT_EQ(1, cc.getCount(StandardCommandType::LOGIN()));

    cc.increment(StandardCommandType::LOGIN());
    ASSERT_EQ(2, cc.getCount(StandardCommandType::LOGIN()));

    ASSERT_EQ(2, cc.getTotal());
    ASSERT_EQ(2, cc.getRecentTotal());
    ASSERT_EQ(2, cc.getRecentExactTotal());
}

void testMixedCommandsCount()
{
    Timer::setTime("20070101.010101");

    // 5 second 'recent' time.
    CommandCounter cc(5000);
    ASSERT_EQ(0, cc.getCount(StandardCommandType::POLL()));
    ASSERT_EQ(0, cc.getCount(StandardCommandType::LOGIN()));
    ASSERT_EQ(0, cc.getCount(StandardCommandType::CHECK()));
    ASSERT_EQ(0, cc.getCount(StandardCommandType::LOGOUT()));
    cc.increment(StandardCommandType::POLL());
    cc.increment(StandardCommandType::LOGIN());
    cc.increment(StandardCommandType::CHECK());
    cc.increment(StandardCommandType::LOGOUT());
    ASSERT_EQ(1, cc.getCount(StandardCommandType::POLL()));
    ASSERT_EQ(1, cc.getCount(StandardCommandType::LOGIN()));
    ASSERT_EQ(1, cc.getCount(StandardCommandType::CHECK()));
    ASSERT_EQ(1, cc.getCount(StandardCommandType::LOGOUT()));

    ASSERT_EQ(4, cc.getTotal());
    ASSERT_EQ(4, cc.getRecentTotal());
    ASSERT_EQ(4, cc.getRecentExactTotal());

    // 10 seconds later, which is greater than 5 seconds.
    Timer::setTime("20070101.010111");
    ASSERT_EQ(4, cc.getTotal());
    ASSERT_EQ(4, cc.getRecentTotal());

    // Alters cached recent total val.
    ASSERT_EQ(0, cc.getRecentExactTotal());
    ASSERT_EQ(0, cc.getRecentTotal());
    ASSERT_EQ(4, cc.getTotal());
}

void testPollCount()
{
    Timer::setTime("");
    Timer::setTime("20070101.010101");
    // 5 second recent window.
    CommandCounter cc(5000);

    ASSERT_EQ(0, cc.getRecentCount(StandardCommandType::POLL()));

    // event at 1sec
    cc.increment(StandardCommandType::POLL());
    ASSERT_EQ(1, cc.getRecentCount(StandardCommandType::POLL()));

    // advance to 3sec
    Timer::setTime("20070101.010103");
    cc.increment(StandardCommandType::POLL());
    ASSERT_EQ(2, cc.getRecentCount(StandardCommandType::POLL()));

    // advance to 8sec, (1sec event gone)
    Timer::setTime("20070101.010108");
    ASSERT_EQ(1, cc.getRecentCount(StandardCommandType::POLL()));

    // advance to 17sec (7 second event gone).
    Timer::setTime("20070101.010117");
    ASSERT_EQ(0, cc.getRecentCount(StandardCommandType::POLL()));
}

int main(int argc, char* argv[])
{
	init("etc/toolkit2.conf");

	TEST_run(testLoginCount);
	TEST_run(testMixedCommandsCount);
	TEST_run(testPollCount);

	return TEST_errorCount();
}

File: src/test/unit/java/com/ausregistry/jtoolkit2/session/CommandCounterTest.java
package com.ausregistry.jtoolkit2.session;

import static org.junit.Assert.assertEquals;

import org.junit.Before;
import org.junit.Test;

import com.ausregistry.jtoolkit2.se.StandardCommandType;
import com.ausregistry.jtoolkit2.Timer;

/**
 * @author anthony
 *
 */
public class CommandCounterTest {
    private CommandCounter counter;

    @Before
    public void setUp() throws Exception {
        Timer.setTime("20070101.010101");
        Timer.setTime("20070101.010101");
        counter = new CommandCounter(5000);
        counter.increment(StandardCommandType.LOGIN);
        counter.increment(StandardCommandType.CHECK);
        counter.increment(StandardCommandType.LOGOUT);
    }

    /**
     * Test method for {@link com.ausregistry.jtoolkit2.session.CommandCounter#increment(com.ausregistry.jtoolkit2.se.CommandType)}.
     */
    @Test
    public void testIncrement() {
        assertEquals(Long.valueOf(0L), Long.valueOf(counter.getCount(StandardCommandType.POLL)));
        counter.increment(StandardCommandType.POLL);
        assertEquals(Long.valueOf(1L), Long.valueOf(counter.getCount(StandardCommandType.POLL)));
    }

    /**
     * Test method for {@link com.ausregistry.jtoolkit2.session.CommandCounter#getCount(com.ausregistry.jtoolkit2.se.CommandType)}.
     */
    @Test
    public void testGetCount() {
        assertEquals(Long.valueOf(1L), Long.valueOf(counter.getCount(StandardCommandType.LOGIN)));
        assertEquals(Long.valueOf(1L), Long.valueOf(counter.getCount(StandardCommandType.CHECK)));
        assertEquals(Long.valueOf(1L), Long.valueOf(counter.getCount(StandardCommandType.LOGOUT)));
    }

    /**
     * Test method for {@link com.ausregistry.jtoolkit2.session.CommandCounter#getRecentCount(com.ausregistry.jtoolkit2.se.CommandType)}.
     */
    @Test
    public void testGetRecentCount() {
        assertEquals(Long.valueOf(1L), Long.valueOf(counter.getRecentCount(StandardCommandType.LOGIN)));
        assertEquals(Long.valueOf(1L), Long.valueOf(counter.getRecentCount(StandardCommandType.CHECK)));
        assertEquals(Long.valueOf(1L), Long.valueOf(counter.getRecentCount(StandardCommandType.LOGOUT)));
    }

    /**
     * Test method for {@link com.ausregistry.jtoolkit2.session.CommandCounter#getTotal()}.
     */
    @Test
    public void testGetTotal() {
        assertEquals(Long.valueOf(3L), Long.valueOf(counter.getTotal()));
    }

    /**
     * Test method for {@link com.ausregistry.jtoolkit2.session.CommandCounter#getRecentTotal()}.
     */
    @Test
    public void testGetRecentTotal() {
        assertEquals(Long.valueOf(3L), Long.valueOf(counter.getRecentTotal()));
        // 1 minute later, recent total should be the same.
        Timer.setTime("20070101.010201");
        assertEquals(Long.valueOf(3L), Long.valueOf(counter.getRecentTotal()));
    }

    /**
     * Test method for {@link com.ausregistry.jtoolkit2.session.CommandCounter#getExactRecentTotal()}.
     */
    @Test
    public void testGetExactRecentTotal() {
        assertEquals(Long.valueOf(3L), Long.valueOf(counter.getExactRecentTotal()));
        // 2 seconds later, exact recent total should be unchanged.
        Timer.setTime("20070101.010103");
        assertEquals(Long.valueOf(3L), Long.valueOf(counter.getExactRecentTotal()));
        // 7 seconds later, all commands are considered old (> 5s).
        Timer.setTime("20070101.010108");
        assertEquals(Long.valueOf(0L), Long.valueOf(counter.getExactRecentTotal()));
        counter.increment(StandardCommandType.CHECK);
        // 1 second later, only the check command is considered recent.
        Timer.setTime("20070101.010109");
        assertEquals(Long.valueOf(1L), Long.valueOf(counter.getExactRecentTotal()));
    }
}

